src/CoreBundle/private/library/classes/dbobjects/CronJobs/TCMSCronJob.class.php

```diff
@@ -123,53 +126,104 @@ private function outputResult($error)
+        $lastPlannedExecution = null;
+        if ('' !== $this->sqlData['last_execution']) {
+            $lastPlannedExecution = \DateTime::createFromFormat(
+                'Y-m-d H:i:s',
+                $this->sqlData['last_execution']
+            );
         }
 
-        $query = 'UPDATE `'.MySqlLegacySupport::getInstance()->real_escape_string($this->table)."`
-                   SET `last_execution` = '".MySqlLegacySupport::getInstance()->real_escape_string(date('Y-m-d H:i:s', $newTime))."',
-                   `real_last_execution` = '".MySqlLegacySupport::getInstance()->real_escape_string(date('Y-m-d H:i:s'))."'
-                 WHERE `id` = '".MySqlLegacySupport::getInstance()->real_escape_string($this->id)."'";
-        MySqlLegacySupport::getInstance()->query($query);
+        $executeEveryNMinutes = (int)$this->sqlData['execute_every_n_minutes'];
 
-        date_default_timezone_set($sServerDateTimeZoneSetting);
+        if (0 === $executeEveryNMinutes) {
+            throw new InvalidArgumentException(
+                sprintf(
+                    'CronJob %s has an invalid value of "%s" for the execute_every_n_minutes field',
+                    $this->sqlData['name'],
+                    $this->sqlData['execute_every_n_minutes']
+                )
+            );
+        }
+
+        return new CronJobScheduleDataModel(
+            $executeEveryNMinutes,
+            (int)$this->sqlData['unlock_after_n_minutes'],
+            '1' === $this->sqlData['lock'],
+            $lastPlannedExecution
+        );
+    }
+
+    /**
+     * updates the last execution time.
+     */
+    protected function _UpdateLastExecutionTime()
+    {
+        $scheduler = $this->getCronJobScheduler();
+        try {
+            $schedule = $this->getSchedule();
+        } catch (InvalidArgumentException $e) {
+            $this->getLogger()->error($e->getMessage(), __FILE__, __LINE__);
+
+            return;
+        }
+        $timeProvider = $this->getTimeProvider();
+        $now = $timeProvider->getDateTime();
+
+        $plannedExecutionTime = $scheduler->calculateCurrentPlannedExecutionDate($schedule);
+
+        $this->getDatabaseConnection()->update(
+            $this->table,
+            [
+                'last_execution'      => $plannedExecutionTime->format('Y-m-d H:i:s'),
+                'real_last_execution' => $now->format('Y-m-d H:i:s'),
+            ],
+            ['id' => $this->id]
+        );
     }
 
     protected function UpdateLastExecutionOnStart()
     {
-        $query = 'UPDATE `'.MySqlLegacySupport::getInstance()->real_escape_string($this->table)."`
-                   SET `last_execution` = '".MySqlLegacySupport::getInstance()->real_escape_string(date('Y-m-d H:i:s'))."'
-                 WHERE `id` = '".MySqlLegacySupport::getInstance()->real_escape_string($this->id)."'";
-        MySqlLegacySupport::getInstance()->query($query);
+        $nowObject = new \DateTime('now');
```

Comment:
-> $now

This should then also use the TimeProvider?

Reply:
k